### DISCUSSION TITLE
Incremental update of parse tree?

### DISCUSSION THREAD

**Discussion by mblinsitu** [tree-sitter/tree-sitter#1954] [Q&A]
The documentation states that when editing the source code, tree_sitter "will create a new tree that internally shares structure with the old tree."

Is there a way to find out which nodes are indeed shared, i.e. which part of the tree did not change?

I am using tree-sitter-js, and when I store the nodes of the old tree in a WeakMap, none of the nodes of the new tree are in the map, indicating that no nodes were shared.

Is this normal? Is it because I am using the JS bindings?

The reason I am interested in this is that I want to provide real-time preview of a Markdown-like language. With incremental updates to the tree, I can incrementally update the DOM, rather than either replace all of it, or figure out what has changed by comparing the trees.

Thanks.

**Comment by ahlinc** [comment]
Currently you can try to use a custom tree-sitter build as described there: https://github.com/ahlinc/tree-sitter/pull/4
![Screenshot from 2022-11-09 05-22-33](https://user-images.githubusercontent.com/14666676/200730495-b0904cfb-336e-4180-a290-a9a0df6a7164.png)
Green dots show nodes that were affected by changes, so they were replaced in the redacted instance of the tree.
